File: Algorithms/merge_sort.rb
def merge_sort array
  return array if array.length == 1
  middle = array.length / 2
  left = array[0...middle]
  right = array[middle...array.length]
  merge( merge_sort(left), merge_sort(right) )
end

def merge a1, a2
  final = []
  until a1.empty? || a2.empty?
    final << ((a1[0] < a2[0]) ? a1.shift : a2.shift)
  end
  final + a1 + a2
end

p merge_sort [4,2,5,65,2,5,4,2,4]

File: Algorithms/merge_sort.py
def merge_sort(unsorted):
    if len(unsorted) <= 1:
        return unsorted
    mid = len(unsorted) / 2
    left = unsorted[0:mid]
    right = unsorted[mid:]
    return merge(merge_sort(left), merge_sort(right))

def merge(list_a, list_b):
    result = []
    while list_a and list_b:
        if list_a[0] > list_b[0]:
            result.append(list_b[0])
            list_b = list_b[1:]
        else:
            result.append(list_a[0])
            list_a = list_a[1:]
    return result + list_a + list_b

